Included all the mld calculations into the branch. Also the binning a…

tillmrtz opened the pull request:
…lgorithm and the plot_profile function were taken and updated from the interactive branch.

MOchiara commented:
Nice Till! I will have a better look later but I have two questions now:

- Can the binning function be merged with grid2d in utilities? I feel like they should kind of do the same things or I am not understanding and checked too quickly the functions?
- Shall we add the glidertools method I added to `compute_mld()`?
- I think maybe `plot_profile()` could be doing something very similar to `plot_basic_vars()`. Shall I maybe try to remove  `plot_basic_vars()` and work out a way to use `plot_profile()` to have a similar plot?

I am just suggesting this to try and avoid having too many functions that try to do something very similar to each other

MOchiara commented:
Also, isn't the threshold method the same as glidertools but with strict density threshold and reference depth (0.03 and 10m)?

callumrollo commented:
I'm working with Till on debugging test errors on this. Once those are resolved we'll fix the merge conflicts

tillmrtz commented:
> Also, isn't the threshold method the same as glidertools but with strict density threshold and reference depth (0.03 and 10m)?

I now changed my function, so that the threshold calculation works similar to the one from glidertools (i.e. reference depth, threshold and variable could be chosen). Additionally my function enables to apply a specific binning and the density at the reference depth is linearly interpolated instead of taking the closest value

tillmrtz commented:
> Nice Till! I will have a better look later but I have two questions now:
> 
> * Can the binning function be merged with grid2d in utilities? I feel like they should kind of do the same things or I am not understanding and checked too quickly the functions?
> * Shall we add the glidertools method I added to `compute_mld()`?
> * I think maybe `plot_profile()` could be doing something very similar to `plot_basic_vars()`. Shall I maybe try to remove  `plot_basic_vars()` and work out a way to use `plot_profile()` to have a similar plot?
> 
> I am just suggesting this to try and avoid having too many functions that try to do something very similar to each other

1. I have changed the construct2dgrid function, since before the function was just taking the left edge of the bin intervals for both x and y values, even though it was actually binning the data. This is problematic, when your output is actually binned, but to the middle of the inout values (i.e. for depth, actual binning 0-10,10-20,20-30m ..., but the output is not 5,15,25,... but 0,10,20,30). This causes that the last values of the grid are always empty, because the number of bins is N-1. 
Now I added the possibility to choose if the last value of the x and/or y grid is the left edge of the bin (like before), which is helpful if data is already binned evenly (e.g. profile numbers) or if the function is actually used for binning (for not binned data like depth). In that case the bin center is actually more logically to use afterwards.
For making the function work, I had to increase the length of x and y grid first and then decrease its size at the end. But this way, the grid has no nan because of missing bins. This is not problematic, because the right edge of the binning interval is not taken into account.

2. compute_mld_glidertools is obsolete now
3. We might discuss what the best choice is here

callumrollo commented:
@tillmrtz is this ready for review and merge or do you want to make any further changes

tillmrtz commented:
> @tillmrtz is this ready for review and merge or do you want to make any further changes

I'll make some changes, so that also the interactive functions work with bin_profile instead of bin_data and then implement one last change to mld computation, where I found errors, when the profile is empty.
Then after resolving potential merge conflicts, the mld stuff should be ready for the final pull
